File: examples/ts/covenant-vault/CovenantVault.tsop.ts
import { SmartContract, assert, PubKey, Sig, Addr, ByteString, SigHashPreimage, checkSig, checkPreimage, hash160, extractOutputHash, hash256 } from 'tsop-lang';

class CovenantVault extends SmartContract {
  readonly owner: PubKey;
  readonly recipient: Addr;
  readonly minAmount: bigint;

  constructor(owner: PubKey, recipient: Addr, minAmount: bigint) {
    super(owner, recipient, minAmount);
    this.owner = owner;
    this.recipient = recipient;
    this.minAmount = minAmount;
  }

  public spend(sig: Sig, amount: bigint, txPreimage: SigHashPreimage) {
    // Owner must authorize
    assert(checkSig(sig, this.owner));
    assert(checkPreimage(txPreimage));

    // Enforce minimum output amount (covenant rule)
    assert(amount >= this.minAmount);
  }
}

File: examples/go/covenant-vault/CovenantVault.tsop.go
package contract

import "tsop"

type CovenantVault struct {
	tsop.SmartContract
	Owner     tsop.PubKey `tsop:"readonly"`
	Recipient tsop.Addr   `tsop:"readonly"`
	MinAmount tsop.Bigint `tsop:"readonly"`
}

func (c *CovenantVault) Spend(sig tsop.Sig, amount tsop.Bigint, txPreimage tsop.SigHashPreimage) {
	// Owner must authorize
	tsop.Assert(tsop.CheckSig(sig, c.Owner))
	tsop.Assert(tsop.CheckPreimage(txPreimage))

	// Enforce minimum output amount (covenant rule)
	tsop.Assert(amount >= c.MinAmount)
}
